qiskit/transpiler/passes/crosstalk_adaptive_schedule.py

```diff
@@ -0,0 +1,717 @@
+                The keys of the crosstalk_prop are tuples for ordered tuples for CX gates
+                e.g., (0, 1) corresponding to CX 0, 1 in the hardware.
+                Each key has an associated value dict which specifies the conditional error rates
+                with nearby gates e.g., (0, 1) : {(2, 3) : 0.2, (2) : 0.15} means that
+                CNOT 0, 1 has an error rate of 0.2 when it is executed in parallel with CNOT 2,3
+                and an error rate of 0.15 when it is executed in parallel with a single qubit
+                gate on qubit 2.
+            weight_factor (float): weight of gate error/crosstalk terms in the objective
+                weight_factor*fidelities + (1-weight_factor)*decoherence errors.
+                Weight can be varied from 0 to 1, with 0 meaning that only decoherence
+                errors are optimized and 1 meaning that only crosstalk errors are optimized.
+                weight_factor should be tuned per application to get the best results.
+            measured_qubits (list): a list of qubits that will be measured in a particular circuit.
+                This arg need not be specified for circuits which already include measure gates.
+                The arg is useful when a subsequent module such as state_tomography_circuits
+                inserts the measure gates. If CrosstalkAdaptiveSchedule is made aware of those
+                measurements, it is included in the optimization.
+        Raises:
+            ImportError: if unable to import z3 solver
+
+        """
+        super().__init__()
+        if not Z3_AVAIL:
+            raise ImportError('z3-solver is required to use CrosstalkAdaptiveSchedule')
+        self.backend_prop = backend_prop
+        self.crosstalk_prop = crosstalk_prop
+        self.gate_id = {}
+        self.bp_u1_err = {}
+        self.bp_u1_dur = {}
+        self.bp_u2_err = {}
+        self.bp_u2_dur = {}
+        self.bp_u3_err = {}
+        self.bp_u3_dur = {}
+        self.bp_cx_err = {}
+        self.bp_cx_dur = {}
+        self.bp_t1_time = {}
+        self.bp_t2_time = {}
+
+        # Z3 variables
+        self.gate_start_time = {}
+        self.gate_duration = {}
+        self.gate_fidelity = {}
+        self.overlap_amounts = {}
+        self.overlap_indicator = {}
+        self.qubit_lifetime = {}
+        self.dag_overlap_set = {}
+        self.xtalk_overlap_set = {}
+        self.opt = Optimize()
+        self.measured_qubits = []
+        self.measure_start = None
+        self.last_gate_on_qubit = None
+        self.first_gate_on_qubit = None
+        self.fidelity_terms = []
+        self.coherence_terms = []
+        self.model = None
+        self.weight_factor = weight_factor
+        if measured_qubits is None:
+            self.input_measured_qubits = []
+        else:
+            self.input_measured_qubits = measured_qubits
+        self.dag = None
```

Comment:
If I move it to run, I get a bunch of pylint errors which ask me to initialize in __init__
qiskit/transpiler/passes/crosstalk_adaptive_schedule.py:126:12: W0201: Attribute 'bp_u1_err' defined outside __init__ (attribute-defined-outside-init)
qiskit/transpiler/passes/crosstalk_adaptive_schedule.py:651:8: W0201: Attribute 'bp_u1_err' defined outside __init__ (attribute-defined-outside-init)
...

One option is to initialize in both init and run. What do you think?

Reply:
I think you're both right -- keeping the variables here is a good way to document all the tools being used, and keeps linter happy (avoids `NoAttribute` error in the future), and we do need to be careful to eliminate side effects of calling `run` twice.

I'd say all the variables that get updated stay as they are in init, and also need to be reset at the start (or end??) of your call to `run`. Maybe a `reset` method is in order?

On a related note, it seems like `parse_backend_properties` should be called in `__init__`. The result of running `parse_backend_properties` will always be the same since the pass is initialized with a backend. Running it in `run` is wasteful if `run` may be called more than once.

Reply:
Sounds good, added a reset function.

Reply:
The issue of stateful passes is a general problem and there are many stateful passes currently: https://github.com/Qiskit/qiskit-terra/issues/2558

We have talked about different ways of resolving this issue, but it is outside the scope of this PR.

I'm fine with the reset here to unblock this PR, but we need a general solution that works everywhere (cc @1ucian0)